cmake/CPM.cmake

```diff
@@ -710,7 +711,10 @@ function(CPMAddPackage)
     string(TOLOWER ${CPM_ARGS_NAME} lower_case_name)
     set(origin_parameters ${CPM_ARGS_UNPARSED_ARGUMENTS})
     list(SORT origin_parameters)
-    if(CPM_USE_NAMED_CACHE_DIRECTORIES)
+    if(CPM_ARGS_CUSTOM_CACHE_DIR)
+      # Application set a custom unique directory name
+      set(download_directory ${CPM_SOURCE_CACHE}/${lower_case_name}/${CPM_ARGS_CUSTOM_CACHE_DIR})
```

Comment:
Version must be handled manually (like with a combination of the download version and a manual sha of the patch_command files).
The current sha will not work when e.g. patch_command refers to external files, so it is not fool proof either.
Is the description in the readme clear enough?

My use case with some more details:
* Shared toolchain that are several GB and may be used in many application repos. The patch command contains external files adjusting the installation (licenses that may be updated more often than the toolchain).
  * As paths are in the patch_command, the sha will differ and the paths will not be shared
  * As there are external files in the patch_command, the sha is not reliable, requires change to the arguments too.
  * The sha makes it hard to see what is in the cache - I may want to clean certain test versions only
  * Using Docker is inconvenient, the toolchain has a few branches

I would have used SOURCE_DIR for this change if not compatibility had been broken...

Reply:
That's true, patch commands using external files will definitely create issues. I'm still unsure about how I feel about applying patches in the cache as it can create very hard to track issues. In the JavaScript world using `npm` or `yarn`, dependencies are copied from the cache to a project-specific location and any patches are applied there. That way the original cache remains valid no matter how the patch is applied. Maybe a mechanism like that would make sense as well here?

I'm also not completely against this approach, just the name could be confusing as it seems to imply an absolute path. How about calling it something like `CUSTOM_CACHE_KEY` instead?

Reply:
> That's true, patch commands using external files will definitely create issues. I'm still unsure about how I feel about applying patches in the cache as it can create very hard to track issues. In the JavaScript world using `npm` or `yarn`, dependencies are copied from the cache to a project-specific location and any patches are applied there. That way the original cache remains valid no matter how the patch is applied. Maybe a mechanism like that would make sense as well here?

Maybe, even if my several GB packages should not be copied.
That would men to intercept and handle the fetch content arguments.
A bigger feature, I believe this can coexist without maintenance issues.

> I'm also not completely against this approach, just the name could be confusing as it seems to imply an absolute path. How about calling it something like `CUSTOM_CACHE_KEY` instead?

Fine, I pushed a change.